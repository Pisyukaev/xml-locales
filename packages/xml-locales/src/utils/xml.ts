import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import xmlFormatter from 'xml-formatter';
import type {
	X2jOptionsOptional,
	XmlBuilderOptionsOptional
} from 'fast-xml-parser';
import type { XMLFormatterOptions } from 'xml-formatter';

import type { XmlJson } from './types.js';

export interface XmlConstructor {
	parserOptions?: X2jOptionsOptional;
	builderOptions?: XmlBuilderOptionsOptional;
	formateOptions?: XMLFormatterOptions;
}

const defaultParserOptions: X2jOptionsOptional = {
	trimValues: false,
	ignoreDeclaration: true,
	attributeNamePrefix: 'key_',
	alwaysCreateTextNode: true,
	ignoreAttributes: false
};

const defaultBuilderOptions: XmlBuilderOptionsOptional = {
	ignoreAttributes: false,
	attributeNamePrefix: 'key_',
	processEntities: false
};

const defaultFormatterOptions: XMLFormatterOptions = {
	collapseContent: true,
	indentation: '  '
};

const defaultXmlOptions: XmlConstructor = {
	parserOptions: defaultParserOptions,
	builderOptions: defaultBuilderOptions,
	formateOptions: defaultFormatterOptions
};

export class XmlParser {
	private parser: XMLParser;
	private builder: XMLBuilder;
	private formatter: (xmlString: string) => string;

	constructor(xmlOptions?: XmlConstructor) {
		const { parserOptions, builderOptions, formateOptions } = {
			...(xmlOptions || {}),
			...defaultXmlOptions
		};

		this.parser = new XMLParser(parserOptions);
		this.builder = new XMLBuilder(builderOptions);
		this.formatter = (xmlString: string) =>
			xmlFormatter(xmlString, formateOptions);
	}

	xmlToJson(xmlData: string | Buffer): XmlJsonData {
		return new XmlJsonData(this.parser.parse(xmlData));
	}

	jsonToXml(xmlJson: XmlJson): string {
		return this.builder.build(xmlJson);
	}

	formate(xmlString: string): string {
		return this.formatter(xmlString);
	}
}

export class XmlJsonData {
	resources: XmlJson['resources'];

	constructor(xmlJson: XmlJson) {
		this.resources = xmlJson.resources;
	}
}
