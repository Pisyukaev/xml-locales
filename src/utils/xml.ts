import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import xmlFormatter from 'xml-formatter';
import type {
	X2jOptionsOptional,
	XmlBuilderOptionsOptional
} from 'fast-xml-parser';
import type { XMLFormatterOptions } from 'xml-formatter';

interface XmlConstructor {
	parserOptions?: X2jOptionsOptional;
	builderOptions?: XmlBuilderOptionsOptional;
	formateOptions?: XMLFormatterOptions;
}

const defaultParserOptions: X2jOptionsOptional = {
	trimValues: true,
	ignoreDeclaration: true,
	attributeNamePrefix: 'key_',
	alwaysCreateTextNode: true,
	ignoreAttributes: false
};

const defaultBuilderOptions: XmlBuilderOptionsOptional = {
	ignoreAttributes: false,
	attributeNamePrefix: 'key_'
};

const defaultFormatterOptions: XMLFormatterOptions = {
	collapseContent: true,
	indentation: '  '
};

const defaultXmlOptions: XmlConstructor = {
	parserOptions: defaultParserOptions,
	builderOptions: defaultBuilderOptions,
	formateOptions: defaultFormatterOptions
};

export class Xml {
	private parser: XMLParser;
	private builder: XMLBuilder;
	private formatter: (xmlString: string) => string;

	constructor(xmlOptions?: XmlConstructor) {
		const { parserOptions, builderOptions, formateOptions } = {
			...(xmlOptions || {}),
			...defaultXmlOptions
		};

		this.parser = new XMLParser(parserOptions);
		this.builder = new XMLBuilder(builderOptions);
		this.formatter = (xmlString: string) =>
			xmlFormatter(xmlString, formateOptions);
	}

	xmlToJson(xmlData: string): XmlJson {
		return this.parser.parse(xmlData);
	}

	jsonToXml(xmlJson: XmlJson, needFormate: boolean = true): string {
		const xmlString = this.builder.build(xmlJson);

		if (needFormate) {
			return this.formate(xmlString);
		}

		return xmlString;
	}

	formate(xmlString: string): string {
		return this.formatter(xmlString);
	}
}
