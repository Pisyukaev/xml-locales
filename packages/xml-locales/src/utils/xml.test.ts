import { describe, expect, test } from 'vitest';

import {
	checkKeyValueExist,
	checkXmlKey,
	replaceXmlNodeValue
} from './helpers.js';
import { XmlJsonData, XmlParser } from './xml.js';

const xmlData = `
<resources>
  <string name="key1">value1</string>
  <string name="key2">value2</string>
</resources>
`;

describe('XmlParser', () => {
	test('should parse xml to json', () => {
		const xmlParser = new XmlParser();
		const jsonData = xmlParser.xmlToJson(xmlData);
		expect(jsonData).toMatchObject({
			resources: {
				string: [
					{
						key_name: 'key1',
						'#text': 'value1'
					},
					{
						key_name: 'key2',
						'#text': 'value2'
					}
				]
			}
		});
	});

	test('should parse json to xml', () => {
		const xmlParser = new XmlParser();
		const jsonData = xmlParser.xmlToJson(xmlData);
		const xmlString = xmlParser.jsonToXml(jsonData);
		expect(xmlString).toMatchSnapshot();
	});

	test('should format xml string', () => {
		const xmlParser = new XmlParser();
		const jsonData = xmlParser.xmlToJson(xmlData);
		const xmlString = xmlParser.jsonToXml(jsonData);
		const formattedXmlString = xmlParser.formate(xmlString);
		expect(formattedXmlString).toMatchSnapshot();
	});
});

describe('XmlJsonData', () => {
	test('should check if key exist', () => {
		const xmlParser = new XmlParser();
		const jsonData = xmlParser.xmlToJson(xmlData);
		const xmlJsonData = new XmlJsonData(jsonData);

		expect(checkXmlKey(xmlJsonData, 'key1')).toBe(true);
		expect(checkXmlKey(xmlJsonData, 'key3')).toBe(false);
	});

	test('should check if key value exist', () => {
		const xmlParser = new XmlParser();
		const jsonData = xmlParser.xmlToJson(xmlData);
		const xmlJsonData = new XmlJsonData(jsonData);
		expect(checkKeyValueExist('value1', xmlJsonData.resources.string)).toBe(
			true
		);
		expect(checkKeyValueExist('value3', xmlJsonData.resources.string)).toBe(
			false
		);
	});

	test('should replace the value of the key', () => {
		const xmlParser = new XmlParser();
		const jsonData = xmlParser.xmlToJson(xmlData);
		const xmlJsonData = new XmlJsonData(jsonData);

		expect(
			checkKeyValueExist(
				'value3',
				replaceXmlNodeValue(xmlJsonData, 'key1', 'value3').resources.string
			)
		).toBe(true);
	});
});
