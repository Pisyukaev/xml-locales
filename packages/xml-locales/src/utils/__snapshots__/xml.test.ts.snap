// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XmlParser > should format xml string 1`] = `
"<resources>
  <string name=\\"key1\\">value1</string>
  <string name=\\"key2\\">value2</string>
</resources>"
`;

exports[`XmlParser > should parse json to xml 1`] = `
"<resources><string name=\\"key1\\">value1</string><string name=\\"key2\\">value2</string>
  
  
</resources>"
`;
